File: frontend/src/types/project.ts
export type AnalysisDepth = 'fast' | 'balanced' | 'deep';
export type GraphBackend = 'sqlite' | 'in_memory' | 'memgraph' | 'falkordb' | 'neo4j';
export type ConstructionMode = 'native_only' | 'native_plus_model_refinement';
export type ProviderType = 'ollama_native_api' | 'openwebui_api' | 'openai_compatible_api';
export type AnalysisProvider = 'native' | 'ollama' | 'openwebui';
export type RemoteSendPolicy =
  | 'graph_metadata_only'
  | 'graph_metadata_plus_selected_snippets'
  | 'full_selected_file_context_only_on_demand';

export interface ModelConfig {
  provider_type: ProviderType;
  base_url: string;
  api_key?: string | null;
  remote_send_policy: RemoteSendPolicy;
  planner_model?: string | null;
  code_model?: string | null;
  query_model?: string | null;
}

export interface ProjectConfig {
  entry_points: string[];
  focus_modules: string[];
  exclude_patterns: string[];
  analysis_depth: AnalysisDepth;
  graph_backend: GraphBackend;
  construction_mode: ConstructionMode;
  model_config_data?: ModelConfig | null;
}

export interface Project {
  id: string;
  name: string;
  path: string;
  language: string;
  source_type?: 'local_path' | 'github_url';
  source_url?: string | null;
  resolved_path?: string | null;
  status: 'pending' | 'indexing' | 'ready' | 'error' | 'refining' | 'cancelled';
  config: ProjectConfig;
  created_at: string;
  last_indexed?: string | null;
  error_message?: string | null;
  file_count: number;
  node_count: number;
  edge_count: number;
}

export interface IndexRequest {
  name: string;
  path: string;
  language: string;
  config: ProjectConfig;
}

export interface ProgressEvent {
  project_id: string;
  stage: string;
  message: string;
  progress: number;
  files_processed: number;
  files_total: number;
  error?: string | null;
}

export interface BrowseFile {
  path: string;
  relative_path: string;
  name: string;
}

File: backend/models/project.py
from __future__ import annotations
from enum import Enum
from typing import Optional, Any
from pydantic import BaseModel
import uuid
from datetime import datetime
from config import (
    OLLAMA_API_KEY,
    OLLAMA_BASE_URL,
    OPENAI_COMPATIBLE_API_KEY,
    OPENAI_COMPATIBLE_BASE_URL,
    OPENWEBUI_API_KEY,
    OPENWEBUI_BASE_URL,
)


class AnalysisDepth(str, Enum):
    FAST = "fast"
    BALANCED = "balanced"
    DEEP = "deep"


class GraphBackend(str, Enum):
    IN_MEMORY = "in_memory"
    SQLITE = "sqlite"
    MEMGRAPH = "memgraph"
    FALKORDB = "falkordb"
    NEO4J = "neo4j"


class ConstructionMode(str, Enum):
    NATIVE_ONLY = "native_only"
    NATIVE_PLUS_MODEL_REFINEMENT = "native_plus_model_refinement"


class ModelProviderType(str, Enum):
    OLLAMA_NATIVE = "ollama_native_api"
    OPENWEBUI = "openwebui_api"
    OPENAI_COMPATIBLE = "openai_compatible_api"


class RemoteSendPolicy(str, Enum):
    GRAPH_METADATA_ONLY = "graph_metadata_only"
    GRAPH_METADATA_PLUS_SNIPPETS = "graph_metadata_plus_selected_snippets"
    FULL_SELECTED_FILE_ON_DEMAND = "full_selected_file_context_only_on_demand"


class ModelConfig(BaseModel):
    provider_type: ModelProviderType = ModelProviderType.OLLAMA_NATIVE
    base_url: str = OLLAMA_BASE_URL
    api_key: Optional[str] = OLLAMA_API_KEY or None
    remote_send_policy: RemoteSendPolicy = RemoteSendPolicy.GRAPH_METADATA_PLUS_SNIPPETS
    planner_model: Optional[str] = None
    code_model: Optional[str] = None
    query_model: Optional[str] = None


def build_default_model_config(provider_type: ModelProviderType) -> ModelConfig:
    if provider_type == ModelProviderType.OPENWEBUI:
        return ModelConfig(
            provider_type=provider_type,
            base_url=OPENWEBUI_BASE_URL,
            api_key=OPENWEBUI_API_KEY or None,
        )
    if provider_type == ModelProviderType.OPENAI_COMPATIBLE:
        return ModelConfig(
            provider_type=provider_type,
            base_url=OPENAI_COMPATIBLE_BASE_URL,
            api_key=OPENAI_COMPATIBLE_API_KEY or None,
        )
    return ModelConfig(
        provider_type=provider_type,
        base_url=OLLAMA_BASE_URL,
        api_key=OLLAMA_API_KEY or None,
    )


class ProjectConfig(BaseModel):
    entry_points: list[str] = []
    focus_modules: list[str] = []
    exclude_patterns: list[str] = []
    analysis_depth: AnalysisDepth = AnalysisDepth.BALANCED
    graph_backend: GraphBackend = GraphBackend.SQLITE
    construction_mode: ConstructionMode = ConstructionMode.NATIVE_ONLY
    model_config_data: Optional[ModelConfig] = None


class ProjectStatus(str, Enum):
    PENDING = "pending"
    INDEXING = "indexing"
    READY = "ready"
    ERROR = "error"
    REFINING = "refining"
    CANCELLED = "cancelled"


class ProjectSourceType(str, Enum):
    LOCAL_PATH = "local_path"
    GITHUB_URL = "github_url"


class Project(BaseModel):
    id: str = ""
    name: str
    path: str
    language: str = "python"
    source_type: ProjectSourceType = ProjectSourceType.LOCAL_PATH
    source_url: Optional[str] = None
    resolved_path: Optional[str] = None
    status: ProjectStatus = ProjectStatus.PENDING
    config: ProjectConfig = ProjectConfig()
    created_at: str = ""
    last_indexed: Optional[str] = None
    error_message: Optional[str] = None
    file_count: int = 0
    node_count: int = 0
    edge_count: int = 0

    def model_post_init(self, __context: Any) -> None:
        if not self.id:
            self.id = str(uuid.uuid4())
        if not self.created_at:
            self.created_at = datetime.utcnow().isoformat()


class IndexRequest(BaseModel):
    name: str
    path: str
    language: str = "python"
    config: ProjectConfig = ProjectConfig()


class ProgressEvent(BaseModel):
    project_id: str
    stage: str
    message: str
    progress: float = 0.0
    files_processed: int = 0
    files_total: int = 0
    error: Optional[str] = None
